fdbserver/VersionedBTree.actor.cpp

```diff
@@ -1286,108 +1289,241 @@ int nextPowerOf2(uint32_t x) {
 
-	void clear() {
-		memset(this, 0, sizeof(RedwoodMetrics));
-		for (auto& level : levels) {
-			level = {};
+		eventReasonsArray(){clear();}
+		void clear(){
+			for(size_t i = 0; i<(size_t)events::MAXEVENTS; i++){
+				for(size_t j = 0; j<(size_t)pagerEventReasons::MAXEVENTREASONS; j++){
+					eventReasons[i][j] = 0;
+				}
+			}
 		}
-		startTime = g_network ? now() : 0;
-	}
+		void addEventReason(events e, pagerEventReasons r){
+			eventReasons[(size_t)e][(size_t)r] += 1;
+		}
+		const unsigned int& getEventReason(events e, pagerEventReasons r){
+			return eventReasons[(size_t)e][(size_t)r];
+		}
+
+		std::string ouputSummary(int currLevel){
+			std::string result = "";
+			std::map<events, std::string> allEvents = {{events::pagerCacheLookup, "pagerCacheLookup"}, {events::pagerCacheHit, "pagerCacheHit"}, {events::pagerCacheMiss, "pagerCacheMiss"}, {events::pagerWrite, "pagerWrite"}};
+			std::map<pagerEventReasons, std::string> allReasons = {{pagerEventReasons::pointRead, "pointRead"}, {pagerEventReasons::rangeRead, "rangeRead"}, {pagerEventReasons::rangePrefetch, "rangePrefetch"}, {pagerEventReasons::commit, "commit"}, {pagerEventReasons::lazyClear, "lazyClear"}, {pagerEventReasons::metaData, "metaData"}};
+			
+			for(auto e = allEvents.begin(); e != allEvents.end(); e++){
+				result += e->second + "\n\t";
+				for(auto r = allReasons.begin(); r != allReasons.end(); r++){
+					std::string num = std::to_string(eventReasons[(size_t)e->first][(size_t)r->first]);
+					result += r->second; 
+					result.append(16 - r->second.length(), ' ');
+					result.append(8 - num.length(), ' ');
+					result += num;
+					result.append(13, ' ');
+				}
+				result +="\n";
+			}
+			return result;
+		}
+		void reportTrace(TraceEvent* t, int h){
+			const std::vector<std::pair<events,pagerEventReasons>> possibleEventReasonPairs = {
+				{events::pagerCacheLookup, pagerEventReasons::pointRead},{events::pagerCacheLookup, pagerEventReasons::rangeRead},
+				{events::pagerCacheLookup, pagerEventReasons::rangePrefetch},{events::pagerCacheLookup, pagerEventReasons::commit},
+				{events::pagerCacheLookup, pagerEventReasons::lazyClear},{events::pagerCacheLookup, pagerEventReasons::metaData},
 
+				{events::pagerCacheHit, pagerEventReasons::pointRead},{events::pagerCacheHit, pagerEventReasons::rangeRead},
+				{events::pagerCacheHit, pagerEventReasons::rangePrefetch},
+				{events::pagerCacheHit, pagerEventReasons::lazyClear},{events::pagerCacheHit, pagerEventReasons::metaData},
+
+				{events::pagerCacheMiss, pagerEventReasons::pointRead},{events::pagerCacheMiss, pagerEventReasons::rangeRead},
+				{events::pagerCacheMiss, pagerEventReasons::rangePrefetch},
+				{events::pagerCacheMiss, pagerEventReasons::lazyClear},{events::pagerCacheMiss, pagerEventReasons::metaData},
+
+				{events::pagerWrite, pagerEventReasons::commit},
+				{events::pagerWrite, pagerEventReasons::lazyClear},{events::pagerWrite, pagerEventReasons::metaData},
+			};
```

Comment:
Although it's tricky to model here, range read prefetch actually only applies to the Leaf level.  Perhaps we should make range read prefetch a non-level-specific metric instead - so move it to the outer metrics struct instead of `Level::Metrics`.  I doubt we would prefetch non-leaf nodes.